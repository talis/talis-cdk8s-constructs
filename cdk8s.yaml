language: typescript
imports:
  - k8s@1.30.0
