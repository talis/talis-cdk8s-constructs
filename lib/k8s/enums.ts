/**
 * Image pull policy.
 *
 * @see https://kubernetes.io/docs/concepts/containers/images#updating-images
 * @default "Always" if :latest tag is specified, or "IfNotPresent" otherwise.
 */
export enum ContainerImagePullPolicy {
  /**
   * Always means that kubelet always attempts to pull the latest image.
   * Container will fail If the pull fails.
   */
  ALWAYS = "Always",
  /**
   * IfNotPresent means that kubelet pulls if the image isn't present on disk.
   * Container will fail if the image isn't present and the pull fails.
   */
  IF_NOT_PRESENT = "IfNotPresent",
  /**
   * Never means that kubelet never pulls an image, but only uses a local image.
   * Container will fail if the image isn't present.
   */
  NEVER = "Never",
}

/**
 * DNS policy for the pod.
 * @default "ClusterFirst".
 */
export enum DNSPolicy {
  /**
   * Any DNS query that does not match the configured cluster domain suffix,
   * such as "www.kubernetes.io", is forwarded to an upstream nameserver by
   * the DNS server. Cluster administrators may have extra stub-domain and
   * upstream DNS servers configured.
   */
  CLUSTER_FIRST = "ClusterFirst",
  /**
   * For Pods running with hostNetwork, you should explicitly set its DNS
   * policy to "ClusterFirstWithHostNet". Otherwise, Pods running with
   * hostNetwork and "ClusterFirst" will fallback to the behavior of
   * the "Default" policy.
   */
  CLUSTER_FIRST_WITH_HOST_NET = "ClusterFirstWithHostNet",
  /**
   * The Pod inherits the name resolution configuration from the node that
   * the Pods run on.
   */
  Default = "Default",
  /**
   * It allows a Pod to ignore DNS settings from the Kubernetes environment.
   * All DNS settings are supposed to be provided using the dnsConfig field
   * in the Pod Spec.
   */
  None = "None",
}

/**
 * Policy for preempting pods with lower priority.
 * @default "PreemptLowerPriority"
 */
export enum PreemptionPolicy {
  /**
   * Pods will be placed in the scheduling queue ahead of lower-priority pods,
   * but they cannot preempt other pods. Non-preempting pods may still be
   * preempted by other, high-priority pods.
   */
  NEVER = "Never",
  /**
   * Allow pods to preempt lower-priority pods (default behavior).
   */
  PREEMPT_LOWER_PRIORITY = "PreemptLowerPriority",
}

/**
 * Restart policy for all containers within the pod.
 *
 * @see https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy
 * @default "Always"
 */
export enum PodSpecRestartPolicy {
  /** Always */
  ALWAYS = "Always",
  /** Never */
  NEVER = "Never",
  /** OnFailure */
  ON_FAILURE = "OnFailure",
}

/**
 * Type of daemon set update.
 *
 * @default "RollingUpdate"
 */
export enum DaemonSetUpdateStrategyType {
  /** OnDelete - Replace the old daemons only when it's killed */
  ON_DELETE = "OnDelete",
  /** RollingUpdate - Replace the old daemons by new ones using rolling update i.e replace them on each node one after the other. */
  ROLLING_UPDATE = "RollingUpdate",
}

/**
 * Type of deployment.
 *
 * @default "RollingUpdate"
 */
export enum DeploymentStrategyType {
  /** Recreate - Kill all existing pods before creating new ones. */
  RECREATE = "Recreate",
  /** RollingUpdate - Replace the old ReplicaSets by new one using rolling update i.e gradually scale down the old ReplicaSets and scale up the new one. */
  ROLLING_UPDATE = "RollingUpdate",
}

/**
 * Type of update strategy.
 *
 * @default "RollingUpdate"
 */
export enum StatefulSetUpdateStrategyType {
  /** OnDelete - triggers the legacy behavior. Version tracking and ordered rolling restarts are disabled. Pods are recreated from the StatefulSetSpec when they are manually deleted. When a scale operation is performed with this strategy,specification version indicated by the StatefulSet's currentRevision.*/
  ON_DELETE = "OnDelete",
  /** RollingUpdate - indicates that update will be applied to all Pods in the StatefulSet with respect to the StatefulSet ordering constraints. When a scale operation is performed with this strategy, new Pods will be created from the specification version indicated by the StatefulSet's updateRevision.*/
  ROLLING_UPDATE = "RollingUpdate",
}

/**
 * Protocol for port.
 *
 * @default "TCP"
 */
export enum PortProtocol {
  /** The SCTP protocol. */
  SCTP = "SCTP",
  /** The TCP protocol. */
  TCP = "TCP",
  /** The UDP protocol. */
  UDP = "UDP",
}

/**
 * Determines how the Service is exposed. Defaults to ClusterIP.
 *
 * @see https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types
 * @default "ClusterIP"
 */
export enum ServiceSpecType {
  /**
   * ClusterIP means a service will only be accessible inside the cluster,
   * via the cluster IP.
   */
  CLUSTER_IP = "ClusterIP",
  /**
   * ExternalName means a service consists of only a reference to an external
   * name that kubedns or equivalent will return as a CNAME record, with no
   * exposing or proxying of any pods involved.
   */
  EXTERNAL_NAME = "ExternalName",
  /**
   * LoadBalancer means a service will be exposed via an external load balancer
   * (if the cloud provider supports it), in addition to 'NodePort' type.
   */
  LOAD_BALANCER = "LoadBalancer",
  /**
   * NodePort means a service will be exposed on one port of every node,
   * in addition to 'ClusterIP' type.
   */
  NODE_PORT = "NodePort",
}
