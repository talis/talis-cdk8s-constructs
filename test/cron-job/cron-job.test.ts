import { Chart, Testing } from "cdk8s";
import { Quantity } from "../../imports/k8s";
import {
  ContainerImagePullPolicy,
  CronJob,
  CronJobProps,
  PodSpecRestartPolicy,
} from "../../lib";
import { makeChart } from "../test-util";

const requiredProps: CronJobProps = {
  schedule: "0 0 13 * 5",
  image: "talis/app:worker-v1",
  release: "v1",
  resources: {
    requests: {
      cpu: Quantity.fromString("100m"),
      memory: Quantity.fromString("100Mi"),
    },
  },
  backoffLimit: 2,
  restartPolicy: PodSpecRestartPolicy.ON_FAILURE,
};

function synthCronJob(
  props: CronJobProps = requiredProps,
  chartLabels: { [key: string]: string } = {},
) {
  const chart = makeChart({
    namespace: "test",
    labels: chartLabels,
  });
  new CronJob(chart, "cron-job-test", props);
  const results = Testing.synth(chart);
  return results;
}

describe("CronJob", () => {
  describe("Props", () => {
    test("Minimal required props", () => {
      const chart = Testing.chart();
      new CronJob(chart, "cron-job-test", requiredProps);
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();
    });

    test("All the props", () => {
      const app = Testing.app();
      const chart = new Chart(app, "test", {
        namespace: "test",
        labels: {
          app: "my-app",
          environment: "test",
          region: "local",
        },
      });
      const selectorLabels = {
        app: "my-app",
        role: "cronjob",
        instance: "test",
      };
      const allProps: Required<Omit<CronJobProps, "makeAffinity">> = {
        ...requiredProps,
        selectorLabels,
        suspend: false,
        suspendJob: true,
        containerName: "my-container",
        workingDir: "/some/path",
        command: ["/bin/sh", "-c", "echo hello"],
        args: ["--foo", "bar"],
        backoffLimit: 1,
        activeDeadlineSeconds: 120,
        ttlSecondsAfterFinished: 30,
        terminationGracePeriodSeconds: 60,
        safeToEvict: false,
        safeToEvictLocalVolumes: ["tmp-dir"],
        env: [{ name: "FOO", value: "bar" }],
        envFrom: [{ configMapRef: { name: "foo-config" } }],
        automountServiceAccountToken: false,
        imagePullPolicy: ContainerImagePullPolicy.ALWAYS,
        imagePullSecrets: [{ name: "foo-secret" }],
        priorityClassName: "high-priority-nonpreempting",
        startingDeadlineSeconds: 200,
        successfulJobsHistoryLimit: 4,
        failedJobsHistoryLimit: 2,
        containers: [{ name: "secondary", image: "second-image" }],
        affinity: {
          podAffinity: {
            requiredDuringSchedulingIgnoredDuringExecution: [
              {
                labelSelector: {
                  matchExpressions: [
                    {
                      key: "app.kubernetes.io/component",
                      operator: "In",
                      values: ["redis"],
                    },
                  ],
                },
                topologyKey: "kubernetes.io/hostname",
              },
            ],
          },
        },
        resources: {
          requests: {
            cpu: Quantity.fromNumber(0.1),
            memory: Quantity.fromString("100Mi"),
          },
          limits: {
            cpu: Quantity.fromNumber(1),
            memory: Quantity.fromString("1Gi"),
          },
        },
        securityContext: {
          runAsUser: 1000,
          runAsGroup: 1000,
          runAsNonRoot: true,
        },
        lifecycle: {
          postStart: {
            exec: {
              command: ["/bin/sh", "-c", "echo hello"],
            },
          },
          preStop: {
            exec: {
              command: ["/bin/sh", "-c", "echo goodbye"],
            },
          },
        },
        initContainers: [
          {
            name: "init-container",
            image: "busybox:1.35.0",
            command: ["/bin/sh", "-c", "echo hello"],
          },
        ],
        hostAliases: [
          {
            ip: "127.0.0.1",
            hostnames: ["foo.local"],
          },
        ],
        volumes: [
          {
            name: "tmp-dir",
            emptyDir: {},
          },
        ],
        volumeMounts: [
          {
            name: "tmp-dir",
            mountPath: "/tmp",
          },
        ],
      };
      new CronJob(chart, "cron-job-test", allProps);
      const results = Testing.synth(chart);
      expect(results).toMatchSnapshot();

      const cronJob = results.find((obj) => obj.kind === "CronJob");
      expect(cronJob).toHaveAllProperties(allProps, [
        "containerName",
        "release",
        "selectorLabels",
        "safeToEvict",
        "safeToEvictLocalVolumes",
        "suspendJob",
      ]);
    });

    test("Setting restartPolicy", () => {
      const results = synthCronJob({
        ...requiredProps,
        restartPolicy: PodSpecRestartPolicy.NEVER,
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.spec.restartPolicy",
        "Never",
      );
    });

    test("Setting backoffLimit", () => {
      const results = synthCronJob({
        ...requiredProps,
        backoffLimit: 42,
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty("spec.jobTemplate.spec.backoffLimit", 42);
    });

    describe("Schedule Validation", () => {
      test.each([
        ["Schedule must be specified", "", "Empty schedule"],
        [
          "Invalid cron expression",
          "* * * *",
          "Too short cron schedule string",
        ],
        [
          "Invalid cron expression",
          "* * * * * * *",
          "Too long cron schedule string",
        ],
        [
          "Constraint error, got value 100 expected range 0-59",
          "100 * * * * *",
          "Invalid second",
        ],
        [
          "Constraint error, got value 100 expected range 0-59",
          "* 100 * * * *",
          "Invalid minute",
        ],
      ])(
        "Throws '%s' error with cron schedule (%s) - %s",
        (errorMessage, schedule) => {
          expect(() => {
            synthCronJob({
              ...requiredProps,
              schedule: schedule,
            });
          }).toThrowError(errorMessage);
        },
      );
    });

    test("Setting safeToEvict", () => {
      const results = synthCronJob({
        ...requiredProps,
        safeToEvict: true,
        safeToEvictLocalVolumes: ["foo", "bar"],
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.metadata.annotations",
        {
          "cluster-autoscaler.kubernetes.io/safe-to-evict": "true",
          "cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes":
            "foo,bar",
        },
      );
    });

    test("Allows specifying no affinity", () => {
      const results = synthCronJob({
        ...requiredProps,
        affinity: undefined,
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).not.toHaveProperty(
        "spec.jobTemplate.spec.template.spec.affinity",
      );
    });

    test("Allows specifying custom logic to make affinity", () => {
      const results = synthCronJob({
        ...requiredProps,
        makeAffinity(matchLabels) {
          return {
            podAffinity: {
              requiredDuringSchedulingIgnoredDuringExecution: [
                {
                  labelSelector: {
                    matchExpressions: [
                      {
                        key: "role",
                        operator: "In",
                        values: [matchLabels.role],
                      },
                    ],
                  },
                  topologyKey: "kubernetes.io/hostname",
                },
              ],
            },
          };
        },
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.spec.affinity",
      );
      expect(
        cron.spec.jobTemplate.spec.template.spec.affinity,
      ).toMatchSnapshot();
    });

    test("Allows returning no affinity", () => {
      const results = synthCronJob({
        ...requiredProps,
        makeAffinity() {
          return undefined;
        },
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).not.toHaveProperty(
        "spec.jobTemplate.spec.template.spec.affinity",
      );
    });

    test("selectorLabels can override app", () => {
      const results = synthCronJob({
        ...requiredProps,
        selectorLabels: { app: "foobar" },
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty("metadata.labels.app", "foobar");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.metadata.labels.app",
        "foobar",
      );
    });
  });

  describe("Containers", () => {
    test("Default container name", () => {
      const results = synthCronJob();
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.spec.containers[0].name",
        "app",
      );
    });

    test("Container name from chart's app label", () => {
      const app = Testing.app();
      const chart = new Chart(app, "test", {
        labels: {
          app: "from-chart",
        },
      });
      new CronJob(chart, "worker-test", requiredProps);
      const results = Testing.synth(chart);
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.spec.containers[0].name",
        "from-chart",
      );
    });

    test("Container name from selector label", () => {
      const results = synthCronJob({
        ...requiredProps,
        selectorLabels: { app: "from-selector" },
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.spec.containers[0].name",
        "from-selector",
      );
    });

    test("Container name set explicitly", () => {
      const results = synthCronJob({
        ...requiredProps,
        containerName: "explicit-name",
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty(
        "spec.jobTemplate.spec.template.spec.containers[0].name",
        "explicit-name",
      );
    });

    test("Allows setting multiple containers", () => {
      const results = synthCronJob({
        ...requiredProps,
        containers: [
          {
            name: "sideapp",
            image: "side-image",
          },
        ],
      });
      const cron = results.find((obj) => obj.kind === "CronJob");
      expect(cron).toHaveProperty("spec.jobTemplate");
      const jobTemplate = cron.spec.jobTemplate;
      expect(jobTemplate).toHaveProperty("spec.template.spec.containers");
      expect(jobTemplate.spec.template.spec.containers).toHaveLength(2);
      expect(jobTemplate).toHaveProperty(
        "spec.template.spec.containers[0].name",
        "app",
      );
      expect(jobTemplate).toHaveProperty(
        "spec.template.spec.containers[1].name",
        "sideapp",
      );
    });
  });

  describe("Labels and annotations", () => {
    test("Inherits labels from the chart", () => {
      const results = synthCronJob(requiredProps, {
        app: "my-app",
        environment: "test",
        region: "dev",
      });
      const cron = results.find((obj) => obj.kind === "CronJob");

      expect(cron.metadata.labels).toEqual({
        app: "my-app",
        environment: "test",
        instance: "cron-job-test",
        region: "dev",
        release: "v1",
        role: "cronjob",
      });

      expect(cron.spec.jobTemplate.spec.template.metadata.labels).toEqual({
        app: "my-app",
        environment: "test",
        instance: "cron-job-test",
        region: "dev",
        release: "v1",
        role: "cronjob",
      });
    });

    test("Allows to set custom selectorLabels", () => {
      const results = synthCronJob(
        {
          ...requiredProps,
          selectorLabels: {
            app: "side-app",
            special: "special-value",
          },
        },
        {
          app: "my-app",
          environment: "test",
          region: "dev",
        },
      );
      const cron = results.find((obj) => obj.kind === "CronJob");

      expect(cron.metadata.labels).toEqual({
        app: "side-app",
        environment: "test",
        instance: "cron-job-test",
        region: "dev",
        release: "v1",
        role: "cronjob",
        special: "special-value",
      });

      expect(cron.spec.jobTemplate.spec.template.metadata.labels).toEqual({
        app: "side-app",
        environment: "test",
        instance: "cron-job-test",
        region: "dev",
        release: "v1",
        role: "cronjob",
        special: "special-value",
      });
    });
  });
});
