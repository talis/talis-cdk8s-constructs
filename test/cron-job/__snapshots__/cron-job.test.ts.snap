// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CronJob > Props > All the props 1`] = `
[
  {
    "apiVersion": "batch/v1",
    "kind": "CronJob",
    "metadata": {
      "labels": {
        "app": "my-app",
        "environment": "test",
        "instance": "test",
        "region": "local",
        "release": "v1",
        "role": "cronjob",
      },
      "name": "test-cron-job-test-c88f7970",
      "namespace": "test",
    },
    "spec": {
      "failedJobsHistoryLimit": 2,
      "jobTemplate": {
        "spec": {
          "backoffLimit": 1,
          "template": {
            "metadata": {
              "annotations": {
                "cluster-autoscaler.kubernetes.io/safe-to-evict": "false",
                "cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes": "tmp-dir",
              },
              "labels": {
                "app": "my-app",
                "environment": "test",
                "instance": "test",
                "region": "local",
                "release": "v1",
                "role": "cronjob",
              },
            },
            "spec": {
              "containers": [
                {
                  "args": [
                    "--foo",
                    "bar",
                  ],
                  "command": [
                    "/bin/sh",
                    "-c",
                    "echo hello",
                  ],
                  "env": [
                    {
                      "name": "FOO",
                      "value": "bar",
                    },
                  ],
                  "envFrom": [
                    {
                      "configMapRef": {
                        "name": "foo-config",
                      },
                    },
                  ],
                  "image": "talis/app:worker-v1",
                  "imagePullPolicy": "Always",
                  "lifecycle": {
                    "postStart": {
                      "exec": {
                        "command": [
                          "/bin/sh",
                          "-c",
                          "echo hello",
                        ],
                      },
                    },
                    "preStop": {
                      "exec": {
                        "command": [
                          "/bin/sh",
                          "-c",
                          "echo goodbye",
                        ],
                      },
                    },
                  },
                  "name": "my-container",
                  "resources": {
                    "limits": {
                      "cpu": 1,
                      "memory": "1Gi",
                    },
                    "requests": {
                      "cpu": 0.1,
                      "memory": "100Mi",
                    },
                  },
                  "securityContext": {
                    "runAsGroup": 1000,
                    "runAsNonRoot": true,
                    "runAsUser": 1000,
                  },
                  "volumeMounts": [
                    {
                      "mountPath": "/tmp",
                      "name": "tmp-dir",
                    },
                  ],
                  "workingDir": "/some/path",
                },
              ],
              "imagePullSecrets": [
                {
                  "name": "foo-secret",
                },
              ],
              "initContainers": [
                {
                  "command": [
                    "/bin/sh",
                    "-c",
                    "echo hello",
                  ],
                  "image": "busybox:1.35.0",
                  "name": "init-container",
                },
              ],
              "restartPolicy": "OnFailure",
              "volumes": [
                {
                  "emptyDir": {},
                  "name": "tmp-dir",
                },
              ],
            },
          },
        },
      },
      "schedule": "0 0 13 * 5",
      "startingDeadlineSeconds": 200,
      "successfulJobsHistoryLimit": 4,
      "suspend": false,
    },
  },
]
`;

exports[`CronJob > Props > Minimal required props 1`] = `
[
  {
    "apiVersion": "batch/v1",
    "kind": "CronJob",
    "metadata": {
      "labels": {
        "instance": "cron-job-test",
        "release": "v1",
        "role": "cronjob",
      },
      "name": "test-cron-job-test-c88f7970",
    },
    "spec": {
      "failedJobsHistoryLimit": 1,
      "jobTemplate": {
        "spec": {
          "backoffLimit": 2,
          "template": {
            "metadata": {
              "labels": {
                "instance": "cron-job-test",
                "release": "v1",
                "role": "cronjob",
              },
            },
            "spec": {
              "containers": [
                {
                  "image": "talis/app:worker-v1",
                  "imagePullPolicy": "IfNotPresent",
                  "name": "app",
                  "resources": {
                    "requests": {
                      "cpu": "100m",
                      "memory": "100Mi",
                    },
                  },
                },
              ],
              "restartPolicy": "OnFailure",
            },
          },
        },
      },
      "schedule": "0 0 13 * 5",
      "successfulJobsHistoryLimit": 3,
    },
  },
]
`;
